### PR TITLE
Dual stack bpf - BPF endpoint manager refactor

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description
This PR contains changes to BPF endpoint manager to make it viable to support dual stack in the near future. The changes are listed below
1. **Refactor BPF program loading** - We now load BPF wep, hep programs, attach preamble program, load policy. Have changed this to the following order. loadPrograms->loadPolicy->attach Preamble. This will help me load v4 program, v4 policy, v6 program, v6 policy and then load the preamble for both ingress, egress hooks.
2. **configure attach point** - Since we need attach point to load the preamble, have calculated the attachPoint which is common to both the dataplanes, passed them to v4 or v6 program loading path. 
3. **Add dataplane object to bpf ep mgr** - Introduced dataplane object which is part of bpf endpoint manager and moved the program loading, policy loading functions to the dataplane object.
4. Added v4/v6 policy, filter indices to the interface state.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 1]
```diff
@@ -343,6 +341,45 @@ type bpfEndpointManager struct {
 	natOutIdx int
 
 	arpMap maps.Map
+
+	v4 *bpfEndpointManagerDataplane
+	v6 *bpfEndpointManagerDataplane
+}
+
+type bpfEndpointManagerDataplane struct {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:349]: Perhaps better to think of this like `bpfEpIPFamilyMgr` which manages what is specific for the ip family and it uses the `bpfDataplane` to achieve it?

**Review by tomastigera — COMMENTED** [review]
I would expect that the manager does the top level management and then calls into the per-ip-family manager somehow. I am not sure we need to break the lowest level bpfDataplane interface, but it may need some changes.

Since the manager is atm for a single version, i'd go with whatever it does and extract out anything common rather than adding bunch of `if ipv6Enabled {` That i'd reserve for handling updates and the rest i'd handle with `if m.v6 exist also do the same for v6`

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 2]
```diff
@@ -1795,38 +1935,61 @@ func (m *bpfEndpointManager) doApplyPolicy(ifaceName string) (bpfInterfaceState,
...
+
+	//Attach preamble TC program
+	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		readiness := state.readiness
-		if readiness == ifaceIsReady {
-			if _, err := m.dp.queryClassifier(ifindex, state.qdisc.handle, state.qdisc.prio, false); err != nil {
-				readiness = ifaceNotReady
+		if readiness != ifaceIsReady {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:1962]: you can have this condition outside, for both in/egress

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 3]
```diff
@@ -1856,6 +2033,26 @@ func (m *bpfEndpointManager) applyPolicy(ifaceName string) error {
 	return err
 }
 
+func mergeAttachPoints(ap, ap6 attachPoint) attachPoint {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:2036]:
```suggestion
func mergeAttachPoints(ap4, ap6 attachPoint) attachPoint {
```

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 4]
```diff
@@ -1856,6 +2033,26 @@ func (m *bpfEndpointManager) applyPolicy(ifaceName string) error {
...
+	if aptcV4, v4ok := ap.(*tc.AttachPoint); v4ok {
+		if aptcV6, v6ok := ap6.(*tc.AttachPoint); v6ok {
+			if aptcV4 != nil && aptcV6 == nil {
+				return aptcV4
+			} else if aptcV6 != nil && aptcV4 == nil {
+				return aptcV6
+			}
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:2043]: what if none of them is nil?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 5]
```diff
@@ -1531,29 +1548,82 @@ func (m *bpfEndpointManager) applyProgramsToDirtyDataInterfaces() {
...
+
+			if m.v6 != nil {
+				parallelWG.Add(1)
+				go func() {
+					defer parallelWG.Done()
+					ingressAP6, egressAP6, xdpAP6, err6 = m.v6.applyPolicyToDataIface(iface, hepPtr, &state,
+						tcAttachPoint, xdpAttachPoint)
+				}()
+			} else if m.v4 != nil {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:1573]:
```suggestion
			}
			if m.v4 != nil {
```

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 6]
```diff
@@ -1531,29 +1548,82 @@ func (m *bpfEndpointManager) applyProgramsToDirtyDataInterfaces() {
...
+				ingressAP4, egressAP4, xdpAP4, err4 = m.v4.applyPolicyToDataIface(iface, hepPtr, &state,
+					tcAttachPoint, xdpAttachPoint)
+			}
+
+			parallelWG.Wait()
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:1578]: do you need to wait here? If you do not spin up v4 in goroutine,  it syncs and you can run the whole thin in parallel, no

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 7]
```diff
@@ -2801,7 +3085,11 @@ func (m *bpfEndpointManager) ensureNoProgram(ap attachPoint) error {
 	}
 
 	// Forget the policy debug info
-	m.removePolicyDebugInfo(ap.IfaceName(), 4, ap.HookName())
+	if m.ipv6Enabled {
+		m.removePolicyDebugInfo(ap.IfaceName(), 6, ap.HookName())
+	} else {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:3090]: again, it is not "either or" can be both

**Review by tomastigera — CHANGES_REQUESTED** [review]
I think it is getting where we want to get it, I'd suggest not to only fix the tests so that it compiles, but actually add dual stack tests.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 8]
```diff
@@ -802,49 +882,19 @@ func (m *bpfEndpointManager) updateIfaceStateMap(name string, iface *bpfInterfac
 			flags |= ifstate.FlgReady
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:882]: I will do the readiness part as part of the subsequent PRs.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 9]
```diff
@@ -1691,39 +1761,106 @@ func (m *bpfEndpointManager) updateWEPsInDataplane() {
...
+	return nil
+}
 
+func (m *bpfEndpointManager) wepStateFillJumps(ifaceName string, state *bpfInterfaceState) error {
+	if m.ipv6Enabled {
+		err := m.allocJumpIndicesForWEP(ifaceName, &state.v6)
+		if err != nil {
+			return err
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1843]: This PR does not turn on dual stack yet. Will change all this when enabling dual stack in a subsequent PR

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 10]
```diff
@@ -1691,39 +1761,106 @@ func (m *bpfEndpointManager) updateWEPsInDataplane() {
...
+	return nil
+}
+
+func (m *bpfEndpointManager) dataIfaceStateFillJumps(ifaceName string, state *bpfInterfaceState) error {
+	if m.ipv6Enabled {
+		err := m.allocJumpIndicesForDataIface(ifaceName, &state.v6)
+		if err != nil {
+			return err
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1858]: Same as above.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 4 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:2043]: Will handle this case, when turning on dual stack.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 5 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1573]: Same as above.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 7 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:3090]: Same as above.

**Review by tomastigera — APPROVED** [review]
LGTM lets bank this intermediate step.

**Pull request merged**